scalding-commons-macros/src/main/scala/com/twitter/scalding/commons/macros/impl/ordser/ScroogeOrderedBuf.scala

```diff
@@ -0,0 +1,108 @@
+/*
+ Copyright 2014 Twitter, Inc.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+package com.twitter.scalding.commons.macros.impl.ordser
+
+import scala.language.experimental.macros
+import scala.reflect.macros.Context
+
+import com.twitter.scalding._
+import java.nio.ByteBuffer
+import com.twitter.scalding.serialization.OrderedSerialization
+import com.twitter.bijection.macros.impl.IsCaseClassImpl
+import com.twitter.scrooge.{ ThriftUnion, ThriftStruct }
+import com.twitter.scalding.macros.impl.ordser._
+
+object ScroogeOrderedBuf {
+  def dispatch(c: Context)(buildDispatcher: => PartialFunction[c.Type, TreeOrderedBuf[c.type]]): PartialFunction[c.Type, TreeOrderedBuf[c.type]] = {
+    import c.universe._
+
+    val pf: PartialFunction[c.Type, TreeOrderedBuf[c.type]] = {
+      case tpe if tpe <:< typeOf[ThriftStruct] && !(tpe =:= typeOf[ThriftStruct]) && !(tpe <:< typeOf[ThriftUnion]) => ScroogeOrderedBuf(c)(buildDispatcher, tpe)
+    }
+    pf
+  }
+
+  def apply(c: Context)(buildDispatcher: => PartialFunction[c.Type, TreeOrderedBuf[c.type]], outerType: c.Type): TreeOrderedBuf[c.type] = {
+    import c.universe._
+    def freshT(id: String) = newTermName(c.fresh(id))
+
+    val dispatcher = buildDispatcher
+
+    val companionSymbol = outerType.typeSymbol.companionSymbol
+
+    val fieldNames: List[String] = companionSymbol.asModule.moduleClass.asType.toType
+      .declarations
+      .filter(_.name.decoded.endsWith("Field "))
+      .collect{ case s: TermSymbol => s }
+      .filter(_.isStatic)
+      .filter(_.isVal)
+      .map { t =>
+        val decodedName = t.name.decoded
+        val cased = decodedName.take(1).toLowerCase ++ decodedName.drop(1)
+        cased.dropRight(6)
+      }.toList
+
+    val elementData: List[(c.universe.Type, TermName, TreeOrderedBuf[c.type])] =
+      outerType
+        .declarations
+        .collect { case m: MethodSymbol => m }
+        .filter(m => fieldNames.contains(m.name.toTermName.toString))
+        .map { accessorMethod =>
+          val fieldType = accessorMethod.returnType.asSeenFrom(outerType, outerType.typeSymbol.asClass)
+          val b: TreeOrderedBuf[c.type] = dispatcher(fieldType)
```

Comment:
Can you add a comment about this being where the recursion occurs? Also, can this pick up implicits set up by the user at this point? I guess not since it is returning a TreeOrderedBuf which is a macro construct. I guess this does not need to get implicits from the user since we know all the possible thrift types. If that is the intent, please add some documentation to explain.